### DISCUSSION TITLE
Apps with SDK 31 not installing on Andriod 12 mobile phones

### DISCUSSION THREAD

**Discussion by ponarangan** [airsdk/Adobe-Runtime-Support#2039] [Q&A]
Hi,My app with SDK version 31 is not installing on mobile phones with Android 12 version. What should i do? Kindly help me
![WhatsApp Image 2022-08-01 at 1 51 05 PM
.

**Comment by ajwfrost** [comment]
Assuming you uploaded via AAB, have you tried installing that locally via ADB? You may get an error message output from that.. bit surprised that it's not installing on Android 12 and yet passed the Play Store checks! Potentially there's some other dependency happening here - is it literally all Android 12 devices that are failing? and all at install-time?

**Reply by ajwfrost** [reply]
... and which AIR SDK did you use? as I've just seen #2040 ...

**Reply by ajwfrost** [reply]
FYI, I can install it on a phone here (Google Pixel 6, Android 12) - but it crashes on start-up.

From the logcat:
```
--------- beginning of crash
08-01 10:28:27.110 16081 16099 E AndroidRuntime: FATAL EXCEPTION: pool-3-thread-1
08-01 10:28:27.110 16081 16099 E AndroidRuntime: Process: air.escapegamestudio.RoomEscapeAmazingAdventure, PID: 16081
08-01 10:28:27.110 16081 16099 E AndroidRuntime: java.lang.IllegalArgumentException: air.escapegamestudio.RoomEscapeAmazingAdventure: Targeting S+ (version 31 and above) requires that one of FLAG_IMMUTABLE or FLAG_MUTABLE be specified when creating a PendingIntent.
08-01 10:28:27.110 16081 16099 E AndroidRuntime: Strongly consider using FLAG_IMMUTABLE, only use FLAG_MUTABLE if some functionality depends on the PendingIntent being mutable, e.g. if it needs to be used with inline replies or bubbles.
08-01 10:28:27.110 16081 16099 E AndroidRuntime: 	at android.app.PendingIntent.checkFlags(PendingIntent.java:375)
```
Not sure what is creating the PendingIntent though..?


Full log is below in case it helps:
```
08-01 10:28:27.075   705   705 D Zygote  : Forked child process 16081
08-01 10:28:27.075  1580  1626 I ActivityManager: Start proc 16081:air.escapegamestudio.RoomEscapeAmazingAdventure/u0a187 for pre-top-activity {air.escapegamestudio.RoomEscapeAmazingAdventure/air.escapegamestudio.RoomEscapeAmazingAdventure.AIRAppEntry}
08-01 10:28:27.082 16081 16081 E mazingAdventur: Not starting debugger since process cannot load the jdwp agent.
08-01 10:28:27.091 16081 16081 D CompatibilityChangeReporter: Compat change id reported: 171979766; UID 10187; state: ENABLED
08-01 10:28:27.095 16081 16081 W mazingAdventur: Entry not found
08-01 10:28:27.095 16081 16081 W mazingAdventur: Entry not found
08-01 10:28:27.095 16081 16081 W mazingAdventur: Entry not found
08-01 10:28:27.096 16081 16081 W mazingAdventur: JIT profile information will not be recorded: profile file does not exist.
08-01 10:28:27.096 16081 16081 W mazingAdventur: JIT profile information will not be recorded: profile file does not exist.
08-01 10:28:27.097 16081 16081 W mazingAdventur: JIT profile information will not be recorded: profile file does not exist.
08-01 10:28:27.100 16081 16081 V GraphicsEnvironment: ANGLE Developer option for 'air.escapegamestudio.RoomEscapeAmazingAdventure' set to: 'default'
08-01 10:28:27.100 16081 16081 V GraphicsEnvironment: ANGLE GameManagerService for air.escapegamestudio.RoomEscapeAmazingAdventure: false
08-01 10:28:27.100 16081 16081 V GraphicsEnvironment: Neither updatable production driver nor prerelease driver is supported.
08-01 10:28:27.100 16081 16081 D NetworkSecurityConfig: No Network Security Config specified, using platform default
08-01 10:28:27.100 16081 16081 D NetworkSecurityConfig: No Network Security Config specified, using platform default
08-01 10:28:27.101 16081 16081 I MultiDex: VM with version 2.1.0 has multidex support
08-01 10:28:27.101 16081 16081 I MultiDex: Installing application
08-01 10:28:27.101 16081 16081 I MultiDex: VM has multidex support, MultiDex support library is disabled.
08-01 10:28:27.109 16081 16100 D vulkan  : searching for layers in '/data/app/~~8CL8sB25GRCEriMfCtAQSA==/air.escapegamestudio.RoomEscapeAmazingAdventure-ulBH6syxtteuKSs9tTCSYg==/lib/arm64'
08-01 10:28:27.109 16081 16100 D vulkan  : searching for layers in '/data/app/~~8CL8sB25GRCEriMfCtAQSA==/air.escapegamestudio.RoomEscapeAmazingAdventure-ulBH6syxtteuKSs9tTCSYg==/base.apk!/lib/arm64-v8a'
08-01 10:28:27.109 16081 16100 D vulkan  : searching for layers in '/data/app/~~8CL8sB25GRCEriMfCtAQSA==/air.escapegamestudio.RoomEscapeAmazingAdventure-ulBH6syxtteuKSs9tTCSYg==/split_config.arm64_v8a.apk!/lib/arm64-v8a'
08-01 10:28:27.110 16081 16100 D vulkan  : searching for layers in '/data/app/~~8CL8sB25GRCEriMfCtAQSA==/air.escapegamestudio.RoomEscapeAmazingAdventure-ulBH6syxtteuKSs9tTCSYg==/split_config.en.apk!/lib/arm64-v8a'
08-01 10:28:27.110 16081 16100 D vulkan  : searching for layers in '/data/app/~~8CL8sB25GRCEriMfCtAQSA==/air.escapegamestudio.RoomEscapeAmazingAdventure-ulBH6syxtteuKSs9tTCSYg==/split_config.xxhdpi.apk!/lib/arm64-v8a'
08-01 10:28:27.110 16081 16099 D CompatibilityChangeReporter: Compat change id reported: 160794467; UID 10187; state: ENABLED
--------- beginning of crash
08-01 10:28:27.110 16081 16099 E AndroidRuntime: FATAL EXCEPTION: pool-3-thread-1
08-01 10:28:27.110 16081 16099 E AndroidRuntime: Process: air.escapegamestudio.RoomEscapeAmazingAdventure, PID: 16081
08-01 10:28:27.110 16081 16099 E AndroidRuntime: java.lang.IllegalArgumentException: air.escapegamestudio.RoomEscapeAmazingAdventure: Targeting S+ (version 31 and above) requires that one of FLAG_IMMUTABLE or FLAG_MUTABLE be specified when creating a PendingIntent.
08-01 10:28:27.110 16081 16099 E AndroidRuntime: Strongly consider using FLAG_IMMUTABLE, only use FLAG_MUTABLE if some functionality depends on the PendingIntent being mutable, e.g. if it needs to be used with inline replies or bubbles.
08-01 10:28:27.110 16081 16099 E AndroidRuntime: 	at android.app.PendingIntent.checkFlags(PendingIntent.java:375)
08-01 10:28:27.110 16081 16099 E AndroidRuntime: 	at android.app.PendingIntent.getBroadcastAsUser(PendingIntent.java:645)
08-01 10:28:27.110 16081 16099 E AndroidRuntime: 	at android.app.PendingIntent.getBroadcast(PendingIntent.java:632)
08-01 10:28:27.110 16081 16099 E AndroidRuntime: 	at androidx.work.impl.utils.ForceStopRunnable.getPendingIntent(ForceStopRunnable.java:273)
08-01 10:28:27.110 16081 16099 E AndroidRuntime: 	at androidx.work.impl.utils.ForceStopRunnable.isForceStopped(ForceStopRunnable.java:151)
08-01 10:28:27.110 16081 16099 E AndroidRuntime: 	at androidx.work.impl.utils.ForceStopRunnable.forceStopRunnable(ForceStopRunnable.java:171)
08-01 10:28:27.110 16081 16099 E AndroidRuntime: 	at androidx.work.impl.utils.ForceStopRunnable.run(ForceStopRunnable.java:102)
08-01 10:28:27.110 16081 16099 E AndroidRuntime: 	at androidx.work.impl.utils.SerialExecutor$Task.run(SerialExecutor.java:91)
08-01 10:28:27.110 16081 16099 E AndroidRuntime: 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
08-01 10:28:27.110 16081 16099 E AndroidRuntime: 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
08-01 10:28:27.110 16081 16099 E AndroidRuntime: 	at java.lang.Thread.run(Thread.java:920)
08-01 10:28:27.112  1580  4142 W ActivityTaskManager:   Force finishing activity air.escapegamestudio.RoomEscapeAmazingAdventure/.AIRAppEntry
08-01 10:28:27.112  1580 16104 I DropBoxManagerService: add tag=data_app_crash isTagEnabled=true flags=0x2
08-01 10:28:27.112 16081 16100 D mali_cmarp_predictor: checking cmar_predictor for air.escapegamestudio.RoomEscapeAmazingAdventure
08-01 10:28:27.119 16081 16098 V FA      : Deferring to Google Analytics for Firebase for event data collection. https://goo.gl/J1sWQy
```

**Reply by KupoGames** [reply]
I've got the same problem.
My app works on my Android 8 device, and crashes on my Android 12 device due to:
`08-01 10:50:25.207 11009 11035 E AndroidRuntime: java.lang.IllegalArgumentException: air.EpicBattleFantasy5: Targeting S+ (version 31 and above) requires that one of FLAG_IMMUTABLE or FLAG_MUTABLE be specified when creating a PendingIntent.`

I'm using the Distriqt ANEs for Ads, Play Services and In-App purchases, and I assume the issue may be that I haven't updated those in a few months, so I'll need to try that. I tried a simpler app which does not use ANEs, and it installed and ran correctly on Android 12.

I've temporarily fixed the problem by putting this in my XML file:
`<uses-sdk android:minimumSdkVersion="19" android:targetSdkVersion="30" />`
But this is only a short-term solution I guess.

**Reply by ponarangan** [reply]
> ... and which AIR SDK did you use? as I've just seen #2040 ...

Harman SDK 31

**Reply by ponarangan** [reply]
My apps with SDK Version is not opening on Android 12 devices. What could be the issue? The Problem is with the SDK or ANE ?

**Reply by ajwfrost** [reply]
@KupoGames @ponarangan  could we perhaps get the AAB file that you used for uploading to the Play Store here? (or local APK file if that's what you'd done)?
https://transfer.harman.com/requests/Cg2lyAfM3leHJ80vcNHCcJ for upload
thanks

**Reply by ajwfrost** [reply]
Thanks - so yes it looks like there may be some ANEs (androidx, and gms) that may be using `PendingIntent.getActivity()` without explicitly setting the mutability flag...

Can you check with all the ANEs updated to their latest versions?

thanks

![image](https://user-images.githubusercontent.com/6400804/182208905-f8755183-4714-4681-b329-a39e6fd65693.png)

**Reply by KupoGames** [reply]
I'm not sure if I'll have time to update the ANEs this week, but I'll keep you updated when I get around to it.

**Reply by FliplineStudios** [reply]
@KupoGames How long ago did you update the ANEs?  And are you targeting 31 for the Android build?  I think it was around March that all of the Distriqt ANEs were updated for 31 for the target SDK version, previously you had to keep targeting 30 if you were using the ANEs since there were some bigger changes involved in 31 and the ANEs weren't always compatible.  If this is for a live build, for a quick fix you could leave the existing ANEs and target 30 if you're not already explicitly doing this, since that's still acceptable until November.

**Reply by marchbold** [reply]
Hi all, this definitely looks like an issue caused by an old version of some extensions. They have all been updated to resolve this issue when we did the update for android 31 so this shouldn't occur with the latest releases. That particular log seems to reference the `androidx.work` extension, but you'll likely need to update all the extensions if you are still running an old version of the androidx dependencies. 

Good time to consider using apm, would generally be as simple as:

```
apm update
apm generate app-descriptor src/MyApp-app.xml
``` 

(aside from any major code changes).

**Reply by KupoGames** [reply]
@ajwfrost 
I've updated the ANEs, and my app now correctly targets Android 31, and starts up without crashing on Android 12 and Android 8 devices. So the outdated ANEs were definitely the reason for this error message in my case:

`08-01 10:50:25.207 11009 11035 E AndroidRuntime: java.lang.IllegalArgumentException: air.EpicBattleFantasy5: Targeting S+ (version 31 and above) requires that one of FLAG_IMMUTABLE or FLAG_MUTABLE be specified when creating a PendingIntent.`

@marchbold 
I had some trouble setting up apm (it just says "application descriptor not found" when I run it), so I updated the ANEs manually. Everything seems to work as before, except ads do not load - no error is given in ActionScript, and the ads show as available and initialised, so I'm stumped about what I've done wrong. I'm pretty sure I've entered my app ID and package name and other stuff correctly in the XML.

Logcat seems to show some failure configuring the ads, but I don't understand what's going on:

```
08-02 13:57:43.434 28859 28859 D com.distriqt.Adverts: GoogleAdMobRewardedVideoAd::[1778d35c-4293-4fec-aff8-b4f18d856a1e]:onAdFailedToLoad( [3] No ad config. )
08-02 13:57:43.434 28859 28859 D com.distriqt.Adverts: GoogleAdMobRewardedVideoAd::[1778d35c-4293-4fec-aff8-b4f18d856a1e]:onAdFailedToLoad(): adapter: 
08-02 13:57:43.435 28859 28859 I Ads     : Ad failed to load : 3
```

Here's the rest of the logcat if it helps.
[admob fail to load ad.txt](https://github.com/airsdk/Adobe-Runtime-Support/files/9242587/admob.fail.to.load.ad.txt)

Thanks.

**Reply by marchbold** [reply]
Nothing stands out there, looks like a legitimate error from AdMob. Are you using test ad ids or are you attempting to test with live ads? AdMob is very strict about using live ads to test and can ban your account if you do.

**Reply by KupoGames** [reply]
Weird, I guess I'll keep experimenting with it. Thanks for reminding me about the test ads - my device was on the test device list before, but for whatever reason that list was now empty, so I put it back on there.

**Reply by ponarangan** [reply]
I have updated all the new extensions. Still i have get start up error for my apps on android 12 devices.
Room Escape Amazing Adventure

**Reply by ajwfrost** [reply]
I've just installed it on Android 9 and it looks like a missing dependency?
```
08-04 09:28:28.106 21004 21004 E AndroidRuntime: FATAL EXCEPTION: main
08-04 09:28:28.106 21004 21004 E AndroidRuntime: Process: air.escapegamestudio.RoomEscapeAmazingAdventure, PID: 21004
08-04 09:28:28.106 21004 21004 E AndroidRuntime: java.lang.RuntimeException: Unable to get provider androidx.work.impl.WorkManagerInitializer: java.lang.ClassNotFoundException: Didn't find class "androidx.work.impl.WorkManagerInitializer" on path: DexPathList[[zip file "/data/app/air.escapegamestudio.RoomEscapeAmazingAdventure-IaIp4qimN1R-FH1l3seeMg==/base.apk", zip file "/data/app/air.escapegamestudio.RoomEscapeAmazingAdventure-IaIp4qimN1R-FH1l3seeMg==/split_config.arm64_v8a.apk", zip file "/data/app/air.escapegamestudio.RoomEscapeAmazingAdventure-IaIp4qimN1R-FH1l3seeMg==/split_config.da.apk", zip file "/data/app/air.escapegamestudio.RoomEscapeAmazingAdventure-IaIp4qimN1R-FH1l3seeMg==/split_config.de.apk", zip file "/data/app/air.escapegamestudio.RoomEscapeAmazingAdventure-IaIp4qimN1R-FH1l3seeMg==/split_config.en.apk", zip file "/data/app/air.escapegamestudio.RoomEscapeAmazingAdventure-IaIp4qimN1R-FH1l3seeMg==/split_config.xxhdpi.apk"],nativeLibraryDirectories=[/data/app/air.escapegamestudio.RoomEscapeAmazingAdventure-IaIp4qimN1R-FH1l3seeMg==/lib/arm64, /data/app/air.escapegamestudio.RoomEscapeAmazingAdventure-IaIp4qimN1R-FH1l3seeMg==/base.apk!/lib/arm64-v8a, /data/app/air.escapegamestudio.RoomEscapeAmazingAdventure-IaIp4qimN1R-FH1l3seeMg==/split_config.arm64_v8a.apk!/lib/arm64-v8a, /data/app/air.escapegamestudio.RoomEscapeAmazingAdventure-IaIp4qimN1R-FH1l3seeMg==/split_config.da.apk!/lib/arm64-v8a, /data/app/air.escapegamestudio.RoomEscapeAmazingAdventure-IaIp4qimN1R-FH1l3seeMg==/split_config.de.apk!/lib/arm64-v8a, /data/app/air.escapegamestudio.RoomEscapeAmazingAdventure-IaIp4qimN1R-FH1l3seeMg==/split_config.en.apk!/lib/arm64-v8a, /data/app/air.escapegamestudio.RoomEscapeAmazingAdventure-IaIp4qimN1R-FH1l3seeMg==/split_config.xxhdpi.apk!/lib/arm64-v8a, /system/lib64, /system/vendor/lib64]]
0
```

So it can't find `"androidx.work.impl.WorkManagerInitializer"` - not sure if the same issue would be happening on Android 12, can you get the logcat for that?

**Reply by marchbold** [reply]
Sounds like you haven't updated the manifest additions correctly. Are you using apm or are you doing it manually? 

If you are doing it manually I'd suggest you start from a clean manifest and go through your extensions again, when you update from a much older version you'll likely miss a lot of changes in the manifest additions.

**Reply by ponarangan** [reply]
I have used clean manifest and updated all the extensions yet i am getting start up error. Kindly advice
Room Escape Amazing Adventure

**Reply by ponarangan** [reply]
Now we are getting start up error on Android 4.2 (SDK 17), Android 9 (SDK 28) Android 4.2 (SDK 17)

**Reply by ajwfrost** [reply]
Hi

Does it actually work for you locally? i.e. if you generate an APK file and load it onto a local file there, does it start up?

I've just installed from the play store:

```
--------- beginning of crash
08-05 16:42:37.118 30864 30864 E AndroidRuntime: FATAL EXCEPTION: main
08-05 16:42:37.118 30864 30864 E AndroidRuntime: Process: air.escapegamestudio.RoomEscapeAmazingAdventure, PID: 30864
08-05 16:42:37.118 30864 30864 E AndroidRuntime: java.lang.RuntimeException: Unable to get provider androidx.lifecycle.ProcessLifecycleOwnerInitializer: java.lang.ClassNotFoundException: Didn't find class "androidx.lifecycle.ProcessLifecycleOwnerInitializer" on path: DexPathList[[zip file "/data/app/air.escapegamestudio.RoomEscapeAmazingAdventure-ORsSh5F4I8AW_BTuS3aHLw==/base.apk", zip file "/data/app/air.escapegamestudio.RoomEscapeAmazingAdventure-ORsSh5F4I8AW_BTuS3aHLw==/split_config.arm64_v8a.apk", zip file "/data/app/air.escapegamestudio.RoomEscapeAmazingAdventure-ORsSh5F4I8AW_BTuS3aHLw==/split_config.da.apk", zip file "/data/app/air.escapegamestudio.RoomEscapeAmazingAdventure-ORsSh5F4I8AW_BTuS3aHLw==/split_config.de.apk", zip file "/data/app/air.escapegamestudio.RoomEscapeAmazingAdventure-ORsSh5F4I8AW_BTuS3aHLw==/split_config.en.apk", zip file 
```

If you've got the AAB file and can do "Analyse APK" on that, it's worth checking your dex files to see whether that `androidx.lifecycle.ProcessLifecycleOwnerInitializer` really is missing. If so then it may still be that an ANE is missing, or it may be that there's a naming conflict somehow (we rename all .jar files to the ANE name to try to avoid that though..)

Do you have a custom 'application' class?

Also, I'm wondering whether you may be best off to use AIR SDK to create an Android Studio project (`-target android-studio-debug`), and then ensure you can build and run it from that locally on a device via USB before trying to generate the AAB file...

thanks

**Reply by marchbold** [reply]
Hi, If you have a reference to `androidx.lifecycle.ProcessLifecycleOwnerInitializer` you aren't using the latest manifest additions. That is an old initializer that is no longer supported. See this documentation on the change:

https://docs.airnativeextensions.com/docs/application/migrating-to-androidx/#manifest-changes

**Reply by ponarangan** [reply]
> Hi, If you have a reference to `androidx.lifecycle.ProcessLifecycleOwnerInitializer` you aren't using the latest manifest additions. That is an old initializer that is no longer supported. See this documentation on the change:
> 
> https://docs.airnativeextensions.com/docs/application/migrating-to-androidx/#manifest-changes

I am not exactly getting what should be changed. 

we are using Jave jdk 16.0.2

First I made the changes like the one given below and faced the start up error.
```xml
<!-- APPLICATION STATE EVENTS -->
<provider
    android:name="androidx.startup.InitializationProvider"
    android:authorities="APPLICATION_PACKAGE.androidx-startup"
    android:exported="false" >
    <meta-data
        android:name="androidx.lifecycle.ProcessLifecycleInitializer"
        android:value="androidx.startup" />
</provider>
 ```
And then I made the change like this one and had the same issue. Kindly advise.
```xml
<!-- APPLICATION STATE EVENTS -->
<provider
    android:name="androidx.lifecycle.ProcessLifecycleOwnerInitializer"
    android:authorities="APPLICATION_PACKAGE.lifecycle-process"
    android:exported="false"
    android:multiprocess="true" />
```

**Reply by marchbold** [reply]
Can you provide the error with the new manifest additions? Likely that something else in your manifest is also outdated.

**Reply by marchbold** [reply]
What's the error with this manifest? 

Looks like you still have the old WorkManagerInitializer in there. 

You need to setup the new initializationPRovider to do this:

```xml
<provider android:name="androidx.startup.InitializationProvider" android:authorities="air.escapegamestudio.RoomEscapeAmazingAdventure.androidx-startup" android:exported="false">
  <meta-data
      android:name="androidx.lifecycle.ProcessLifecycleInitializer"
      android:value="androidx.startup" />
  <meta-data
      android:name="androidx.work.WorkManagerInitializer"
      android:value="androidx.startup" />
</provider>
```   

and remove `androidx.work.impl.WorkManagerInitializer`
